sdk/servicebus/service-bus/src/core/streamingReceiver.ts

```diff
@@ -434,15 +459,20 @@ export class StreamingReceiver extends MessageReceiver {
    * Starts the receiver by establishing an AMQP session and an AMQP receiver link on the session.
    *
    * @param onMessage - The message handler to receive servicebus messages.
-   * @param onError - The error handler to receive an error that occurs while receivin messages.
+   * @param onError - The error handler to receive an error that occurs while receiving messages.
    */
   subscribe(onMessage: OnMessage, onError: OnError): void {
     throwErrorIfConnectionClosed(this._context);
 
     this._onMessage = onMessage;
     this._onError = onError;
 
-    this._receiverHelper.addCredit(this.maxConcurrentCalls);
+    const emptySlots = numberOfEmptyIncomingSlots(this.link);
+    this._receiverHelper.addCredit(
+      this.receiveMode === "peekLock"
+        ? Math.min(this.maxConcurrentCalls, emptySlots <= 1 ? 0 : emptySlots - 1)
+        : this.maxConcurrentCalls
+    );
```

Comment:
Say maxConcurrentCalls is 100, but only 11 slots are empty. We add 10 credits giving the user a concurrency of 10. If in a few minutes, all messages are getting settled in time, we are still stuck with 10 and never upgrade to 100. Should we have a todo here to improve this?

Reply:
I'm wondering if the case of 11 slots is even possible at this spot.
This method is only ever called right after the link initialization.. which means we'd have 2048 slots always.

The only change this if-check brings is that... 
∀ `maxConcurrentCalls > 2047`, `creditsToAdd` would become 2047.

Reply:
Hmmm.. the case when the subscriber is closed and the new call for subscribe is made without settling the previously received messages would hit this.

Reply:
A simple fix(TODO) would be to keep checking the "maxConcurrentCalls" whenever the credits are being added(whenever a new message is received) and replenish to full "maxConcurrentCalls" if possible/as needed.

Reply:
> the case when the subscriber is closed and the new call for subscribe is made without settling the previously received messages would hit this.

In that case, it would be worth looking into the benefits of re-using the link for the second subscribe call...

In case of batching, it completely makes sense for us to re-use the link for each call, because the suggested usage is to call it multiple times. In case of streaming, the second subscribe call can only be made after stopping the first one or when the first one fails, both of which are a reaction to something happening and not the normal work flow

cc @richardpark-msft @chradek

Reply:
I think it is a completely valid behavior to stop receiving(closing the subscriber) for a period of time and re-using the same link to receive, this gives users the flexibility of settling the messages as needed(meaning.. both the first set of messages and the second set can be settled as per their will). 

It would be simpler to implement the "new-link" per subscribe. But I'm inclined towards keeping the existing behavior and replenishing the credits based on the `maxConcurrentCalls` whenever we add credits.

Reply:
However, I think we can take this particular change out of this PR.

Reply:
Discussed with @richardpark-msft.. 
Previously, my CreditManager.postProcessing would keep waiting and add the credit when a message is settled.
Instead of the while loop and blocking, we are moving it to settleMessage.

And regarding link re-initializing.. we don't want to do it because the user would do .close() if they want to kill the link.